_assets/patches/geth/0016-fix-leveldb-issue.patch

```diff
@@ -1,85 +1,106 @@
+diff --git a/eth/downloader/downloader.go b/eth/downloader/downloader.go
+index 7ede530a9..60ff8dbc6 100644
+--- a/eth/downloader/downloader.go
++++ b/eth/downloader/downloader.go
+@@ -141,6 +141,8 @@ type Downloader struct {
+ 	quitCh   chan struct{} // Quit channel to signal termination
+ 	quitLock sync.RWMutex  // Lock to prevent double closes
+
++	downloads sync.WaitGroup // Keeps track of the currently active downloads
++
+ 	// Testing hooks
+ 	syncInitHook     func(uint64, uint64)  // Method to call upon initiating a new sync run
+ 	bodyFetchHook    func([]*types.Header) // Method to call upon starting a block body fetch
+@@ -398,7 +400,9 @@ func (d *Downloader) synchronise(id string, hash common.Hash, td *big.Int, mode
+ // specified peer and head hash.
+ func (d *Downloader) syncWithPeer(p *peerConnection, hash common.Hash, td *big.Int) (err error) {
+ 	d.mux.Post(StartEvent{})
++	d.downloads.Add(1)
+ 	defer func() {
++		d.downloads.Done()
```

Comment:
I think it's a deadlock. 
https://play.golang.com/p/msvaoFoyC2M
As i see same for https://github.com/ethereum/go-ethereum/pull/16200.
Could you check it?

Reply:
defer with done is a part of syncWithPeer method and wait is called in terminate. it is different from your example

Reply:
Oh! It's separated function. I've missed Terminate method.